tests/unit/test_utils.py

```diff
@@ -17,3 +60,163 @@ def test_cleanse_integer_columns():
+        )
+        fuzzy_checker.fuzzy_assert_proportion(
+            name="test_phonetic_error_values",
+            observed_numerator=(noised == result).sum(),
+            observed_denominator=len(data),
+            target_proportion=pathway_probability,
+            name_additional=f"result {result} from string {string}",
+        )
+
+    # Assert token noise level is correct.
+    if string == "abc":
+        tokens_per_string = 4
+    elif string == "abzc":
+        tokens_per_string = 3
+    elif string == "jkl":
+        tokens_per_string = 0
+
+    any_token_noised = 1 - (1 - token_probability) ** tokens_per_string
+    fuzzy_checker.fuzzy_assert_proportion(
+        name="generate_phonetic_errors",
+        observed_numerator=(noised != data).sum(),
+        observed_denominator=len(data),
+        target_proportion=any_token_noised,
+    )
+
+
+def test_get_index_to_noise(fuzzy_checker: FuzzyChecker):
+    """
+    Tests that the index length we will noise validates to expected noise level
+    """
+
+    df = pd.DataFrame(index=list(range(1000)))
+    noise_level = 0.62
+    chosen_idx = get_index_to_noise(
+        data=df,
+        noise_level=noise_level,
+        randomness_stream=RANDOMNESS0,
+        additional_key="test_get_index_to_noise",
+        is_column_noise=False,
+    )
+    # Assert that the proportion of rows to noise is correct
+    fuzzy_checker.fuzzy_assert_proportion(
+        name="test_get_index_to_noise",
+        observed_numerator=len(chosen_idx),
+        observed_denominator=len(df),
+        target_proportion=noise_level,
+    )
+
+
+def test_vectorized_choice(fuzzy_checker: FuzzyChecker):
+
+    options = ["Supersonics", "Mariners", "Kraken"]
+    num_choices = 50_000
+    choice_weights = [0.40, 0.35, 0.25]
+    picks = vectorized_choice(
+        options=options,
+        n_to_choose=num_choices,
+        randomness_stream=RANDOMNESS0,
+        weights=choice_weights,
+        additional_key="best_seattle_sports_team",
+    )
+    picks = pd.Series(picks)
+
+    assert (picks.isin(options)).all()
+    # Test weights validate
+    for i in range(len(options)):
+        fuzzy_checker.fuzzy_assert_proportion(
+            name="test_vectorized_choice",
+            observed_numerator=(picks == options[i]).sum(),
+            observed_denominator=len(picks),
+            target_proportion=choice_weights[i],
+            name_additional=f"option {options[i]}",
+        )
+
+
+def test_two_d_array_choice(fuzzy_checker: FuzzyChecker):
+
+    best_sports = ["Basketball", "Baseball", "Football"]
+    sports = pd.Series(best_sports * 1000)
+    options = pd.DataFrame(
+        {
+            "option1": [
+                "Supersonics",
+                "Mariners",
+                "Ducks",
+            ],
+            "option2": [
+                "Nuggets",
+                "Dodgers",
+                "Steelers",
+            ],
+            "option3": [
+                "Blazzers",
```

Comment:
I'm not a sport guy but I don't think this is a team?

Reply:
Its the nickname of thePortland team but I changed it now.